src/main/java/edu/indiana/dlib/amppd/service/impl/MediaServiceImpl.java

```diff
@@ -104,6 +107,41 @@ public MediaServiceImpl(AmppdUiPropertyConfig amppduiConfig) {
 		}		
 	}	
 	
+	/**
+	 * @see edu.indiana.dlib.amppd.service.MediaService.getCollectionSupplementPathname(Primaryfile, String, SupplementType)
+	 */
+	@Override
+	public String getSupplementPathname(Primaryfile primaryfile, String name, SupplementType type) {
+		String pathname = null;
+		Supplement supplement = null;
+		List<? extends Supplement> supplements = null;
+		
+		// find the supplements by its name and associated parent's ID
+		switch(type) {
+		case COLLECTION:
+			supplements = collectionSupplementRepository.findByCollectionIdAndName(primaryfile.getItem().getCollection().getId(), name);
+			break;
+		case ITEM:
+			supplements = itemSupplementRepository.findByItemIdAndName(primaryfile.getItem().getId(), name);
+			break;
+		case PRIMARYFILE:
+			supplements = primaryfileSupplementRepository.findByPrimaryfileIdAndName(primaryfile.getId(), name);
+			break;
+		}		
+		
+		// there should be exactly one supplement found, as supplement is unique by name within its parent's scope
+		if (supplements != null && supplements.size() == 1) {
+			supplement = supplements.get(0);
+		}	
+		
+		// if the exact supplement is found, resolve its pathname to the absolute path
+		if (supplement != null) {
+			pathname = fileStorageService.resolve(supplement.getPathname()).toString();
```

Comment:
Can this move to between line 134-135 and remove the null check?

Reply:
Good point. will do! I'm also adding some unit tests.